app/features/worksheet_generator/tools.py

```diff
@@ -0,0 +1,263 @@
+            if self.verbose:
+                logger.error(f"TypeError during response validation: {e}")
+            return False
+
+    def is_valid_dict_values(self, json_dict):
+        """
+        Check if the dictionary values contain special or newline characters.
+        """
+        for key, value in json_dict.items():
+            if isinstance(value, str) and re.search(r'[\n\r]', value):
+                return False
+        return True
+
+    def format_choices(self, choices: Dict[str, str]) -> List[Dict[str, str]]:
+        return [{"key": k, "value": v} for k, v in choices.items()]
+
+    def create_questions(self, num_questions: int = 5) -> dict[str, str] | list[str]:
+        if self.verbose: logger.info(f"Creating {num_questions} questions")
+
+        if num_questions > 10:
+            return {"message": "error", "data": "Number of questions cannot exceed 10"}
+
+        chain = self.execute()
+
+        generated_questions = []
+        attempts = 0
+        max_attempts = num_questions * 5  # Allow for more attempts to generate questions
+        print("len of gen qs", len(generated_questions))
+        while len(generated_questions) < num_questions and attempts < max_attempts:
+            response = chain.invoke(self.topic)
+            if self.verbose:
+                logger.info(f"Generated response attempt {attempts + 1}: {response}")
+
+            # Directly check if the response format is valid
+            if self.validate_response(response):
+                response["choices"] = self.format_choices(response["choices"])
+                generated_questions.append(response)
+                if self.verbose:
+                    logger.info(f"Valid question added: {response}")
+                    logger.info(f"Total generated questions: {len(generated_questions)}")
+            else:
+                if self.verbose:
+                    logger.warning(f"Invalid response format. Attempt {attempts + 1} of {max_attempts}")
+
+            # Move to the next attempt regardless of success to ensure progress
+            attempts += 1
+
+        # Log if fewer questions are generated
+        if len(generated_questions) < num_questions:
+            logger.warning(f"Only generated {len(generated_questions)} out of {num_questions} requested questions")
+
+        if len(generated_questions) == 10:
+            if self.verbose: logger.info(f"10 questions generated, Deleting vectorstore")
+            self.db.delete_collection()
+
+        # Return the list of questions
+        return generated_questions[:num_questions]
+    def validate_qtype_response(self, hint: str, answer) -> bool:
+        if "single sentence" in hint:
+            return 10 <= len(answer.split()) <= 400
+        if "integer" in hint:
+            try:
+                int(answer)
+                return True
+            except ValueError:
+                return False
+
+    def is_response_relevant(self, response: dict) -> bool:
```

Comment:
This proposal is interesting but costful in terms of production resources. Indeed, Google actually provides metrics to accomplish this semantic search process with their embeddings.